Introduce CAN Bus support

rene opened the pull request:
# CAN Bus support on EVE-OS
This PR introduces the support to the CAN Bus on EVE-OS.

## In summary:
* New I/O device types introduced in EVE-API are defined in pillar
* A canbus pillar's library is introduced to control and setup CAN interfaces
* Support to CAN interfaces is added to pillar: It reads configuration from device model and setup Physical and/or Virtual CAN interfaces defined in the model
* CAN interface "passthrough" is implemented for KVM
* The corresponding documentation is provided
* The capability to list physical CAN interfaces is added to spec.sh script

The design document for this implementation can be accessed in the LF Edge's Wiki: https://wiki.lfedge.org/display/EVE/CAN+Bus+support

## Summary of commits:
1d0b110b9  spec.sh: List physical CAN interfaces
d13e4af44 docs: Add CAN Bus documentation
4c9e53b15 kvm: Introduce CAN Bus passthrough
005287f65 pillar: Introduce CAN Bus support
a58276844 pillar: Introduce CAN Bus library

## Regarding Yetus errors:

### Error 1
<s>The following error is also present in several other source files:
```
golangcilint: import 'github.com/vishvananda/netlink' is not allowed from list 'Main' (depguard)
```
Running Yetus locally I found more than 2k incidences of this error throughout the source code. I couldn't find the root cause and how to solve it.
</s>

### Error 2
<s>Regarding the following error:
```
golangcilint: var-naming: don't use ALL_CAPS in Go names; use CamelCase (revive)
revive: don't use ALL_CAPS in Go names; use CamelCase https://revive.run/r#var-naming
```
I won't fix because I just followed the constants defined for the netlink interface in the Linux kernel and in exactly the same way as other similar constants are defined (all in capital letters) in the unix package from golang: https://github.com/Golangltd/leafltd/blob/282493799a89/src/golang.org/x/sys/unix/ztypes_linux_amd64.go#L331
</s>

Edit: Yetus error were fixed and/or properly ignored.

christoph-zededa commented:
There are no tests?

christoph-zededa commented on pkg/debug/spec.sh:
```diff
@@ -528,6 +528,30 @@ __EOT__
+    if [ "$TYPE" != 280 ]; then
+        continue
+    fi
+    ZTYPE=15
```
```suggestion
    ZTYPE="IO_TYPE_..."
```

christoph-zededa commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,881 @@
+
+// CAN Interface, it implements netlink.Link interface
+type CAN struct {
+	LinkAttrs          netlink.LinkAttrs
```
I am not sure, but perhaps it makes more sense to use composition here?

christoph-zededa replied:
resolved - just as suggestion, feel free to ignore

christoph-zededa commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,881 @@
+	IFLA_CAN_TDC_TDCO_MIN = 0x3
+	IFLA_CAN_TDC_TDCO_MAX = 0x4
+	IFLA_CAN_TDC_TDCF_MIN = 0x5
+	IFLA_CAN_TDC_TDCF_MAX = 0x6
```
would be nice to have a comment here what those abbreviations mean, like TDC or TDCF

rene replied:
Add the relevant comment for TDC meaning.

christoph-zededa commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,881 @@
+		case "triple-sampling":
+			cvErr[16] = setCtrlModeFlag(prop, value, &canProps.CtrlMode)
+			changeCtrlMode = true
+		case "one-shot":
```
you parse `propStr` here and then you parse it in `setCtrlModeFlag` again - would be good having a single responsibility only

rene replied:
I agree, the idea is that in SetupCAN all the parameters (passed as strings) are parsed. However, I've broke the logic to set Ctrl. mode to reduce the complexity of SetupCAN. Ctrl. mode represents just a small set of these parameters. My initial plan was to use the flag value itself to determine directly which flag we want to change, so it would be more easy to execute the switch. However, there is no a single flag to identify the CTRLMODE_TDC, so I decided to also use strings in order to no create more constants and increase the complexity of the code. These are suppose to be executed only when configuring the CAN interface, so I believe we can live with a "double" parsing of a small set of strings...

christoph-zededa replied:
we had a discussion yesterday @rene wanted to look at it again

christoph-zededa commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -2818,6 +2819,30 @@ func handlePhysicalIOAdapterListImpl(ctxArg interface{}, key string,
+				}
+				err = canbus.LinkSetUp(vcan)
+				if err != nil {
+					log.Errorf("Failed to enable virtual CAN interface: %s", err)
```
```suggestion
					err := fmt.Errorf("Failed to enable virtual CAN interface: %s", err)
					log.Error(err)
					ib.Error = err
					ib.ErrorTime = time.Now()
```
then the user in the cloud can see directly the error message

christoph-zededa commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -852,6 +882,27 @@ func (ctx KvmContext) CreateDomConfig(domainName string, config types.DomainConf
+			ID         int
+		}{Machine: ctx.devicemodel, IfName: "", HostIfName: "", ID: 0}
+
+		t, _ = template.New("qemuCANBus").Parse(qemuCANBusTemplate)
```
please don't _ errors

rene replied:
Done.

christoph-zededa replied:
resolved

christoph-zededa commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -852,6 +882,27 @@ func (ctx KvmContext) CreateDomConfig(domainName string, config types.DomainConf
+			IfName     string
+			HostIfName string
+			ID         int
+		}{Machine: ctx.devicemodel, IfName: "", HostIfName: "", ID: 0}
```
You can just move this into the for-loop in line 895 then you don't need dummy values, then you can remove the if (line 885) which helps against having the code deeply nested.

rene replied:
I agree reduce the scope of the variable, but I'm not sure if I want to change it, I just followed exactly how it's implemented for other I/O devices just on the code above this excerpt, like serial and PCI assignments... there will be performance improvements? Otherwise I prefer to keep the same pattern for better code readability...

christoph-zededa replied:
resolved

christoph-zededa commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -852,6 +882,27 @@ func (ctx KvmContext) CreateDomConfig(domainName string, config types.DomainConf
+		t, _ = template.New("qemuCANBus").Parse(qemuCANBusTemplate)
+		id := 0
+		for canIf, canHostIf := range canBusAssignments {
+			fmt.Printf("CAN interface %s connected to host CAN %s\n", canIf, canHostIf)
```
```suggestion
			log.Noticef("CAN interface %s connected to host CAN %s\n", canIf, canHostIf)
```

rene replied:
log module is not used in this kvm.go, I prefer to keep as it is in other parts in the code and let this change to another PR....

christoph-zededa replied:
f.e. logrus is used in kvm.go: https://github.com/lf-edge/eve/pull/3802/files#diff-81e96cd930b1704b4214d18d7ebc150d7f17d3622878aaade9b4e1f2daa4a674L800

`fmt.Printf` does not show the filename of this log message, this makes searching logs for messages that come from `kvm.go` harder.

I fear those PRs will not come ...

eriknordmark replied:
I agree using logrus is better than fmt.Printf

rene replied:
Done.

christoph-zededa replied:
resolved

christoph-zededa commented on pkg/pillar/types/assignableadapters.go:
```diff
@@ -98,6 +98,8 @@ type IoBundle struct {
 	Vfs sriov.VFList
 	// Only used in PhyIoNetEthVF
 	VfParams VfInfo
+	// Used for CAN interface properties
```
is it really only can specific?

f.e. in `devmodel.pb.go` I see: "For example in WWAN to which firmware version to load etc"

rene replied:
Actually nobody was using this field, it wasn't considered by pillar so far... but I agree, since it can be used for other purposes, it's better to generalize the comment....

christoph-zededa replied:
resolved

christoph-zededa replied:
resolved

rene commented on pkg/debug/spec.sh:
```diff
@@ -528,6 +528,30 @@ __EOT__
+    if [ "$TYPE" != 280 ]; then
+        continue
+    fi
+    ZTYPE=15
```
Done.

christoph-zededa replied:
resolved

eriknordmark commented:
Regarding:
The following error is also present in several other source files:

golangcilint: import 'github.com/vishvananda/netlink' is not allowed from list 'Main' (depguard)
Running Yetus locally I found more than 2k incidences of this error throughout the source code. I couldn't find the root cause and how to solve it.

Seems like we need a config file per https://github.com/golangci/golangci-lint/issues/3877 to allow imports other than standard packages. Do we not have a config file at all?

eriknordmark commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -800,6 +816,20 @@ func (ctx KvmContext) CreateDomConfig(domainName string, config types.DomainConf
 				serialAssignments = addNoDuplicate(serialAssignments, ib.Serial)
 			}
+			if ib.Type == types.IoLCAN && ib.Ifname != "" {
+				var canIfName string
```
Below you call the same concept canHostIf. I think that name is more clear and it makes sense to use the same variable naming approach.

eriknordmark reviewed (approved):
LGTM but some of @christoph-zededa 's comments might not be have been considered and implemented. Kicking off tests while that happens.

rene commented:
> Regarding: The following error is also present in several other source files:
> 
> golangcilint: import 'github.com/vishvananda/netlink' is not allowed from list 'Main' (depguard) Running Yetus locally I found more than 2k incidences of this error throughout the source code. I couldn't find the root cause and how to solve it.
> 
> Seems like we need a config file per [golangci/golangci-lint#3877](https://github.com/golangci/golangci-lint/issues/3877) to allow imports other than standard packages. Do we not have a config file at all?

AFAIK we don't. It should be the reason, as you pointed out. I will take a look.

I've addressed most of the comments and updated the PR.

eriknordmark reviewed (approved):
@christoph-zededa can you mark those of your comments which @rene and no longer open as resolved.

Kicking off tests.

christoph-zededa commented:
> > There are no tests?
> 
> Tests were made on real hardware with CAN controllers from devices we support. Unfortunately unit-tests are a bit trick here because real CAN controllers supports in general only a set of the parameters that can be set by the library introduced here. For non-supported parameters I've checked the result comparing with returns from iproute2, which were the same. One option is to create a dummy CAN driver kernel module that would accept all the parameters, so we could check the whole data route from netlink IPC to the driver itself... For now I can just compromise with the tests on real hardware, this introduction can unblock those who are needing CAN device access, and I can continue to working on unit tests....

There are two ways:
* mock the interface - which I think is not very useful as there is not much of a logic in the code here
* use https://netmodule-linux.readthedocs.io/en/latest/howto/can.html

christoph-zededa commented:
> @christoph-zededa can you mark those of your comments which @rene and no longer open as resolved.
> 
> Kicking off tests.

I don't have a "Resolve" Button, but I wrote "resolved" for all resolved comments.

christoph-zededa commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -2818,6 +2819,30 @@ func handlePhysicalIOAdapterListImpl(ctxArg interface{}, key string,
+				if err != nil {
+					log.Errorf("Failed to fetch CAN interface: %s", err)
+				}
+				err = canbus.SetupCAN(canIf, ib.Cbattr)
```
```suggestion
				err = canbus.SetupCAN(canIf, ib.Cbattr)
				ib.Error = err
				ib.ErrorTime = time.Now()
```

this reports the error back to the controller and IMO makes it easier for them to find mistakes, like f.e. when they put in a float instead of a uint32.

rene replied:
`ib.Error = err` doesn't work because err is of type error while ib.Error is a string. But I liked the suggestion, I will convert it and also set ib.Error for other cases as well (VCAN interfaces) and functions....

rene replied:
Done.

rene commented:
Updates in this PR:
- Rebase on top of master
- Addressed more comments from @christoph-zededa

christoph-zededa commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -2818,6 +2819,40 @@ func handlePhysicalIOAdapterListImpl(ctxArg interface{}, key string,
+				}
+			} else if ib.Type == types.IoCAN {
+				// Initialize physical CAN device
+				canIf, err := canbus.GetCANLink(ib.Ifname)
+				if err != nil {
+					ib.Error = err.Error()
+					ib.ErrorTime = time.Now()
+					log.Errorf("Failed to fetch CAN interface: %s", err)
+				}
+				err = canbus.SetupCAN(canIf, ib.Cbattr)
+				if err != nil {
+					ib.Error = err.Error()
+					ib.ErrorTime = time.Now()
+					log.Errorf("Failed to setup CAN interface: %s", err)
+				}
+				err = canbus.LinkSetUp(canIf)
```
What happens if `ib.Ifname` is `shenanigans0`. Then `canbus.GetCANLink` returns an error and it is logged with `log.Errorf` (line 2842).
Then `canbus.SetupCAN` returns an error, because `canIf` is nil and the error is logged.
Then `canbus.LinkSetUp` is called with `canIf=nil` 
then calls `netlink.LinkSetup`, 
then calls `pkgHandle.LinkSetUp`,
which calls `link.Attrs()`, but `link` is canIf which would be nil in this case.

Or does `log.Errorf` call exit and whole pillar shuts down?

eriknordmark replied:
I think all of the above log.Errorf  needs to be followed by the appropriate error return or continue based on the context.

rene replied:
`log.Errorf()` is not falal, so it doesn't shuts pillar down. There would be no problem to call `SetupCAN()` with canIf nil, an error would be thrown and caught by the if. But most probably the older error would be overwritten, so I've improved the code by creating two auxiliary functions to do the CAN logic and return at the first error occurrence....

christoph-zededa replied:
It is not just overwriting the error, it is crashing;
you can easily try it out with a test (domainmgr_test.go):
```
func init() {
       logger, log = agentlog.Init(agentName)
}

func FuzzSetupCAN(f *testing.F) {
       f.Fuzz(func(t *testing.T,
               ifname string,
               key1 string,
               val1 string,
               key2 string,
               val2 string,
               key3 string,
               val3 string,
       ) {
               cbattrs := map[string]string{
                       key1: val1,
                       key2: val2,
                       key3: val3,
               }
               ioBundle := types.IoBundle{
                       Ifname: ifname,
                       Cbattr: cbattrs,
               }

               setupCAN(&ioBundle)
       })
}
```
if you remove the `return err`s in `setupCAN` then you see it crashing.

Btw. Did your opinion about tests now change? ;-)

Resolved!

eriknordmark replied:
Or use log.Fatalf if these failures should never happen. But that is dangerous because if they happen you're likely to get into a reboot loop.

rene commented:
> Regarding: The following error is also present in several other source files:
> 
> golangcilint: import 'github.com/vishvananda/netlink' is not allowed from list 'Main' (depguard) Running Yetus locally I found more than 2k incidences of this error throughout the source code. I couldn't find the root cause and how to solve it.
> 
> Seems like we need a config file per [golangci/golangci-lint#3877](https://github.com/golangci/golangci-lint/issues/3877) to allow imports other than standard packages. Do we not have a config file at all?

@eriknordmark , I've opened the PR https://github.com/lf-edge/eve/pull/3813 to fix this error....

rene commented:
Updates in this PR:
- Rebase on top of master
- Address new comment from @christoph-zededa

christoph-zededa commented:
LGTM

eriknordmark commented:
> I don't have a "Resolve" Button, but I wrote "resolved" for all resolved comments.

Odd that the creator of a comment can't mark it as resolved; that's the only person who can tell that the answer was satisfactory IMHO. But the github creators have some other model in mind I guess.

eriknordmark reviewed (approved):
Some more questions.

rene commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,831 @@
+//
+//nolint:all
+const (
+	IFLA_CAN_TDC          = unix.IFLA_CAN_BITRATE_MAX + 1
```
I can disable it with the comment `//nolint:all`, it works if I run the golangclint locally, I didn't figure out why is not working with yetus.... will take another look...

rene replied:
It works :smile:

rene commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,831 @@
+// The following constants are related to TDC (Transmitter Delay
+// Compensation) and were taken from Linux's kernel header:
+// include/uapi/linux/can/netlink.h (version 6.6)
+// PS: All of them are treat as uint32 in netlink packages
```
yes, will fix it, thanks :+1:

rene replied:
Done.

rene commented:
Updates in this PR:
- Fix more Yetus errors

rene commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,853 @@
+// Setup CAN interface parameters, return in case of any error
+// Parameters to be set are marked in the array props in the following order:
+// BitTiming, DataBitTiming, CtrlMode, Termination, Restart, RestartMs, TDC
+func setupCANParams(link *CAN, canProps CAN, props []bool) error {
```
to reduce the complexity of SetupCAN function and make Yetus happy :smile:

christoph-zededa replied:
Yes, but no.

This is not the way to go; please don't use bool arrays with magic numbers.

What about doing the map with funcs instead?

rene replied:
Is not easy to use map in this case due to the different types of the arguments. I would like to avoid create wrapper functions to force the map usage, but I have an alternative approach to avoid the array of bools... let's discuss it tomorrow...

OhmSpectator commented:
Do we have other PRs related to this one? EVE-API?

OhmSpectator commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,853 @@
+}
+
+// Set a bit flag
+func setBitFlag(flags uint32, mask uint32) uint32 {
```
I usually tried to move such simple functions to the `pkg/pillar/utils` dir.

rene replied:
I agree, I won't migrate in this case because I want to let this library 100% independent of pillar in order to merge it to netlink package and push it upstream in the future....

eriknordmark commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -2818,6 +2819,18 @@ func handlePhysicalIOAdapterListImpl(ctxArg interface{}, key string,
+			} else if ib.Type == types.IoCAN {
+				// Initialize physical CAN device
+				err := setupCAN(ib)
+				if err != nil {
+					log.Errorf("setupCAN failed")
```
Ditto

rene replied:
Same as above.

eriknordmark commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -3309,6 +3325,51 @@ func removeUSBfromKernel() bool {
+func setupVCAN(ib *types.IoBundle) error {
+	vcan, err := canbus.AddVCANLink(ib.Ifname)
+	if err != nil {
+		ib.Error = err.Error()
+		ib.ErrorTime = time.Now()
+		log.Errorf("Failed to add Virtual CAN interface: %s", err)
```
See above - these 3 lines can be removed and done in caller.

milan-zededa commented on docs/CANBus.md:
```diff
@@ -0,0 +1,391 @@
+# CAN Bus support
```
Note that the first sentence [here](https://github.com/lf-edge/eve/blob/master/docs/DEVICE-CONNECTIVITY.md#physical-network-ports) mentions CAN. However, since CAN is not really a network interface, it probably should not be there. Please remove it from there if you agree.

rene replied:
Thanks for pointed out @milan-zededa . It's done.

rene commented:
> First off, huge kudos for diving into adding CAN support – that's really cool and super valuable for the project! 😊 I've been checking out your PR, and it's clear you've put a ton of work into it. I've got a few thoughts and was hoping we could chat about them a bit.
> 
> From what I understand, the current approach introduces CAN-related code directly into the Pillar service. Given Pillar's role as primarily an orchestration service, I wonder if embedding low-level device handling directly into it might lead us away from its core responsibilities. Would it be possible to encapsulate this functionality into a dedicated package that Pillar could then utilize? This could help keep Pillar's codebase more focused and modular.
> 
> Alternatively, considering existing tools like can-utils could offer a way to leverage community-maintained solutions without reinventing the wheel. I recognize that relying on external tools introduces its challenges, especially around integrating and managing interactions. However, the benefits of such an approach, including reducing the complexity within Pillar and potentially increasing our project's maintainability, seem worth discussing.
> 
> My concern is primarily about the long-term sustainability and maintainability of our codebase. We've observed with the network subsystem that having a very narrow maintenance corridor can create bottlenecks. It might be prudent to explore strategies that avoid centralizing too much specialized knowledge within a small group, thus increasing our bus factor.
> 
> Of course, there are significant advantages to developing in-house solutions, particularly in terms of control and adaptability to our specific needs. Yet, I believe finding a balance that allows us to benefit from existing solutions while maintaining the flexibility we require is key.
> 
> This brings me to a broader question that I think is worth discussing as a team: When should we opt to reuse existing tools, and when is it preferable to implement from scratch? And if we decide to implement, where should this code live to best maintain the modular and maintainable nature of our project?
> 
> I hope these thoughts are taken in the collaborative spirit they're intended. Just wanted to say again how much I appreciate the effort you've put into this – it's really impressive work! I'm not suggesting we redo anything that's already been done; just thought it'd be great to chat about how we tackle similar challenges in the future.
> 
> Summoning @eriknordmark, @rouming, @milan-zededa, @christoph-zededa, and whoever is interested =)

@OhmSpectator , all your questions are 1000% valid for discussion. In respect to CAN library introduced in this PR, my very first attempt was to look at netlink's Go package in order to see if it fully supports our needs to manage CAN interfaces. Unfortunately, it does not. CAN management was introduced in beta version but still it doesn't support all the parameters that we want to support on EVE. The approach chosen for implementation it was to use the netlink kernel IPC to manage CAN interfaces. This is exactly how iproute, netlink and other tools do to control network interfaces. In the near future my plan is to port this library to netlink's package, since we also use this package to do network management.

Regarding what we should use on pillar, well, that's a broader discussion, outside the scope of this PR. So I suggest that we create an issue or drop these questions on our Slack channel so we can discuss with the whole EVE team. IMO run tools under the hood are more error prone and makes harder to parse arguments, input/output, receive asynchronous events.... However, I do agree that for some resources we can use dedicated daemons to control as long as we have a good IPC mechanism to talk to them. A very good example of a nice approach: ModemManager.... that's way better to delegate the complexity of modem management to the daemon service and just talk to this service using dbus, libraries and/or another good IPC. That's what @milan-zededa did. But in summary, pillar controls critical and many system resources, that inevitable to face and handle (at some point) low level code.

Again, thanks your comments and questions...

milan-zededa commented on docs/CANBus.md:
```diff
@@ -0,0 +1,391 @@
+        "logicallabel": "can0",
+        "assigngrp" : "",
+        "phyaddrs" : {
+            "ifname" : "can0"
```
Is reference by interface name reliable for CAN devices? Or could the init order of CAN devices and thus their interface names change between reboots?

rene replied:
yes, it is. In case of CAN there is no other possible association, like MAC addresses, serial numbers... perhaps we could look some other attribute for physical controllers, but definitely not for virtual CAN, so the ifname is the common one for all interfaces. Also, now that we have udev a eventual problem could be solved with udev rules, right?

milan-zededa commented on docs/CANBus.md:
```diff
@@ -0,0 +1,391 @@
+        "ztype": "IO_TYPE_VCAN",
+        "phylabel": "vcan0",
+        "logicallabel": "vcan0",
+        "assigngrp" : "",
```
We should probably also document the meaning of `assigngrp` for CAN devices (e.g. that it should be different for LCANs even if they reference the same CAN).
Is it actually looked at by domaingr at all in the case of CAN devices?

rene replied:
Yes, good point. I've updated the documentation.

milan-zededa commented on pkg/debug/spec.sh:
```diff
@@ -530,6 +530,30 @@ __EOT__
+     "ztype": ${ZTYPE},
+     "phylabel": "${IFNAME}",
+     "phyaddrs": {
+       "Ifname": "${IFNAME}"
```
Can CAN device have PCI (or USB) address? Should we include them here as well?

rene replied:
depending on the controller Yes, but QEMU just use the ifname to do the "passt-through", which as explained in the docs, is not a real pass-through, but an emulation.... I have nothing to do with PCI and USB addresses in this case, so that's why they are omitted. For ARM devices usually the controllers are implemented in the SoC, so they don't have PCI or USB address.

milan-zededa commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -800,6 +816,20 @@ func (ctx KvmContext) CreateDomConfig(domainName string, config types.DomainConf
+					canIfName = ib.Logicallabel
+				}
+				if canIfName != "" {
+					logrus.Infof("Adding CAN interface <%s>\n", canIfName)
```
Newline character is not needed in logrus methods.

milan-zededa reviewed (approved):
Nice work! Added just few comments.

rene commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -2818,6 +2819,18 @@ func handlePhysicalIOAdapterListImpl(ctxArg interface{}, key string,
+				// Initialize (create and enable) Virtual CAN device
+				err := setupVCAN(ib)
+				if err != nil {
+					log.Errorf("setupVCAN failed")
```
the error is already logged and handled in setupVCAN() and setupCAN() functions. I don't want to return at this point because they are not fatal, so I think it's safe to not print the message "setupVCAN failed". I agree, it's useless since the error description will be already informed at this point...

rene commented:
> Do we have other PRs related to this one? EVE-API?

All changes for EVE-API and Kernel were already merged. The initial changes for EVE-API are here: https://github.com/lf-edge/eve-api/commit/5dda59a3d85efea7439b97406161eb097ed7e532 (they were updated by @christoph-zededa to add USB devices)

rene commented:
Updates in the PR:
- Addressed latest comments.

Big kudos to @christoph-zededa which proposed the changes now implemented in SetupCAN() in order to remove the switch/case for checking properties, reducing the complexity of the code.

christoph-zededa commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,894 @@
+			return err
+		},
+	}
+	// Set BiTiming
```
```suggestion

	// Set BitTiming
```

eriknordmark commented:
@OhmSpectator thanks for your very thoughful comments.
We definitely want to think about what we can leverage particularly when it makes our lives easier and less complex by leveraging a community which is handling the complexity. Modem manager is an example of that.

For CANbus the need is much simpler - be able to configure the CAN interfaces.
Ideally that should be done using something like the netlink package, but as Renê said that isn't there yet. We should consider switching to using it once it gets more support (or contribute the missing pieces once it is out of beta).
The can-utils do not seem to focus on the configuration but instead have lots of other useful tools to test/debug and use canbus from scripts etc. 

The general question about both how we leverage existing pieces but also how we plug them into EVE is something we should talk more explicitly about (as well as figuring out which existing pieces we can replace and/or refactor.)

eriknordmark commented:
Build failed with
#19 5.072 ERROR: https://dl-cdn.alpinelinux.org/alpine/v3.16/main: temporary error (try again later)
#19 5.072 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.16/main: No such file or directory
#19 10.08 ERROR: https://dl-cdn.alpinelinux.org/alpine/v3.16/community: temporary error (try again later)
#19 10.08 WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.16/community: No such file or directory
so I restarted it.

eriknordmark reviewed (approved):
Re-run eden

rene commented:
> Re-run eden

ok @eriknordmark , in any case I will push an update for this PR... there is an small thing to fix....

uncleDecart reviewed (approved):
LGTM, thanks @rene for this!

I have one general non-critical comment regarding canbus.go module: it looks like you have a group of functions which take link *CAN as first parameter, which makes it feel like a class method. For example 
```go
func SetTermination(link *CAN, termination uint16) error;
```
turns into 
```go
func (link *CAN) SetTermination(termination uint16) error;
```
So you can write this:
```go
link.SetTermination(canProps.Termination)
```

rene commented on pkg/pillar/canbus/canbus.go:
```diff
@@ -0,0 +1,894 @@
+		return nil, err
+	}
+	if nlink.Type() != LinkCAN && nlink.Type() != LinkVCAN {
+		return nil, errors.New("Not a CAN interface")
```
Indeed, that's a better error message, will do it!

rene commented:
> LGTM, thanks @rene for this!
> 
> I have one general non-critical comment regarding canbus.go module: it looks like you have a group of functions which take link *CAN as first parameter, which makes it feel like a class method. For example
> 
> ```go
> func SetTermination(link *CAN, termination uint16) error;
> ```
> 
> turns into
> 
> ```go
> func (link *CAN) SetTermination(termination uint16) error;
> ```
> 
> So you can write this:
> 
> ```go
> link.SetTermination(canProps.Termination)
> ```

@uncleDecart , the reason to not use a receiver in those functions it's mainly because I followed the same approach as in the netlink's Go package functions.... my plan for the future is to merge canbus library into netlink and push it upstream....

rene commented:
Updates in this PR:
* Rebase on top of master
* Addressed @uncleDecart 's comment
* Rewrote SetupCAN: Although @christoph-zededa 's proposal had reduced the complexity, Yetus was still complaining about the complexity of SetupCAN, it was giving false positive for duplicated lines, and the function was indeed still complex... so at the end I rewrote this function with the following improvements:
    - Validation of properties passed in order to setup CAN
    - Removal of maps to different functions, now just a few auxiliary functions are needed
    - Incremental setup, i.e., as soon as we have all properties for a given parameter, setup the parameter in the CAN interface. For instance, as soon as we have all BitTiming properties, setup this parameter in the CAN interface
* Just return at the first error

eriknordmark reviewed (approved):
Re-run eden

eriknordmark reviewed (approved):
Run again
